extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/runtime/ConcurrencyLimiter.java

```diff
@@ -0,0 +1,102 @@
+package io.quarkus.websockets.next.runtime;
+
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.websockets.next.WebSocketServerConnection;
+import io.smallrye.mutiny.helpers.queues.Queues;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+
+/**
+ * Used to limit concurrent invocations.
+ */
+class ConcurrencyLimiter {
+
+    private static final Logger LOG = Logger.getLogger(ConcurrencyLimiter.class);
+
+    private final WebSocketServerConnection connection;
+    private final Queue<Action> queue;
+    private final AtomicLong uncompleted;
```

Comment:
If we can have many limiters, would be nice to sligthly reduce the footprint by using `AtomicXXFieldUpdater`s and inline the counters.

Reply:
Yes, we will have a limiter per client connection by default. I'll create a separate issue to collect various perfomance-related topics.

Reply:
https://github.com/quarkusio/quarkus/issues/39148